src/raftstore/store/worker/region.rs

```diff
@@ -305,24 +451,240 @@ impl Runnable<Task> for Runner {
                 region_id,
                 notifier,
             } => {
-                // It safe for now to handle generating and applying snapshot concurrently,
+                // It is safe for now to handle generating and applying snapshot concurrently,
                 // but it may not when merge is implemented.
                 let ctx = self.ctx.clone();
                 self.pool
                     .execute(move |_| ctx.handle_gen(region_id, notifier))
             }
-            Task::Apply { region_id, status } => self.ctx.handle_apply(region_id, status),
+            Task::Apply { region_id, status } => {
+                self.ctx
+                    .handle_apply(&mut self.pending_delete_ranges, region_id, status)
```

Comment:
IMO, if you return the applied region info here, you can keep all `pending_delete_ranges` related logic inside this `Runner`.